draft-davidson-pp-protocol.md

```diff
@@ -815,223 +589,105 @@ and is necessary for achieving the unforgeability requirement.
-            s_update: s_update
-         }
-4. Output cli_cfg
-~~~
+Note that we must run the verifiable version of the protocol in
+{{I-D.irtf-cfrg-voprf}}. Therefore the `Issuer` takes the role of the
+`Server` running in `modeVerifiable`. In other words, the `Issuer` runs
+`(ctxtI, pkI) = SetupVerifiableServer(suite)`; where `suite` is one of
+the ciphersuites in {{voprf-ciph-recs}}, `ctxt` contains the internal
+VOPRF Server functionality and secret key `skI`, and `pkI` is the Issuer
+public key. Likewise, run `ctxtC = SetupVerifiableClient(suite)` to
+generate the Client context.
+
+## Functionality {#voprf-api}
+
+For the explicit signatures of each of the functions, refer to
+{{pp-api}}.
 
 ### Generate
 
 ~~~
-1. ciph = cli_cfg.s.ciphersuite
-2. GG = ciph.group()
-3. c_data = []
-4. i_data = []
-5. g_data = []
-6. for i in 0..m:
-       1. c_data[i] = GG.scalar_field().random().as_bytes()
-7. (blinds,group_elems) = ciph.VerifiableBlind(c_data)
-8. for i in 0..m:
-       1. i_data[i] = group_elems[i].as_bytes()
-       2. g_data[i] = blinds[i].as_bytes()
-9. Output ClientIssuanceInput {
-              ClientIssuanceProcessing {
-                  client_data: c_data,
-                  gen_data: g_data,
-              },
-              ClientIssuanceElement {
-                  msg_data: i_data,
-              }
-          }
+def Generate(m):
+  inputs = []
+  for i in range(m):
+    inputs[i] = random_bytes()
+  (tokens, blindedTokens) = Blind(inputs)
+  return IssuanceInput {
+           internal: tokens,
+           msg: blindedTokens,
+         }
 ~~~
 
 ### Issue
 
 ~~~
-1. ciph = s_cfg.ciphersuite
-2. pk = s_cfg.pub_key.as_element()
-3. GG = ciph.group()
-4. m = msg_data.length
-5. if m > max_evals: panic(ERR_MAX_EVALS)
-6. G = GG.generator()
-7. elts = msg_data.as_elements();
-8. Z,D = ciph.VerifiableEval(key.as_scalar(),G,pk,elts)
-9. evals = []
-10. for i in 0..m:
-    1.  eval[i] = ServerEvaluation {
-                      data: Z[i].as_bytes();
-                  }
-11. proof = ServerProof {
-                data: D.as_bytes()
-            }
-12. Output IssuanceResponse {
-                evaluations: eval,
-                proof: proof,
-            }
+def Issue(pkI, skI, msg):
+  Ev = Evaluate(skI, pkI, msg)
+  return IssuanceResponse {
+           tokens: Ev.elements,
+           proof: Ev.proof,
+         }
 ~~~
 
 ### Process
 
 ~~~
-1. ciph = cli_cfg.s.ciphersuite
-2. GG = ciph.group()
-3. G = GG.generator()
-4. pk = cli_cfg.s.pub_key.as_element()
-5. M = i_data.as_elements()
-6. Z = evals.as_elements()
-7. r = g_data.as_scalars()
-8. N = ciph.VerifiableUnblind(G,pk,M,Z,r,proof)
-9. if N == "error": panic(ERR_PROOF_VALIDATION)
-10. tokens = []
-11. for i in 0..m:
-        1. issued = N[i].as_bytes()
-        2. rt = RedemptionToken { data: c_data[i], issued: issued }
-        3. tokens[i] = rt
-12. Output tokens
+Process(pkI, input, resp):
+  unblindedTokens = Unblind(pkI, input.data, input.msg, resp)
```

Comment:
Ditto here for the batch interface.

Reply:
Right, but if we have a batched proof then we'll need to evaluate it over all the evaluated elements at once. I don't think this is will be as easy as pushing it all into the for loop?